### ISSUE TITLE
privload_early_inject assigns to global variable before program is relocated

### ISSUE THREAD

**Issue by egrimley** [DynamoRIO/dynamorio#1865]
The function relocate_dynamorio has this comment:

    /* XXX: This routine is called before dynamorio relocation when we are in a
     * fragile state and thus no globals access or use of ASSERT/LOG/STATS!
     */

However, in privload_early_inject, just before where relocate_dynamorio is called, there is:

    kernel_init_sp = (void *)sp;

That's an assignment to a global, and in my experimental branch it causes a segfault because the address of kernel_init_sp is taken from the unrelocated GOT.

Moving the assignment to after the call to relocate_dynamorio is unlikely to be a robust solution as that function is in the same file so the compiler would be free to reorder those things anyway.

You could perhaps make copious use of "volatile" or put relocate_dynamorio in a separate file and hope that there will be no whole-program optimisation, but I would guess that the safest thing to do would be to call relocate_dynamorio from the assembler start-up code. Does that sound reasonable?

**Comment by derekbruening** [comment]
This is a recent regression.  It is not simple to make a test for this that runs on standard distros (we would have to have a whole separate build of libdynamorio on which we did a setarch call or sthg), and so one was not added with #1708.  When I try to reproduce your crash right now, by manually disabling ASLR, I do indeed get libdynamorio at its non-preferred location but I do not get a crash for some reason (though I expect one).

Calling relocate_dynamorio from the _start asm code makes a lot of sense and makes it less likely for a future regression as well.

Given that you are hitting issues (here, and #1864) that just do not show up on any of our tests (which is seemingly not simple to rectify), it may be best for you to run the suite on your setup in a cron job and post the results to our CDash server, as I believe we suggested over email a while back.  The devs list would then get emails on failures.
